src/Bicep.Core/SemanticModel/Compilation.cs

```diff
@@ -59,28 +65,47 @@ private SemanticModel GetSemanticModelInternal()
 
             // bind identifiers to declarations
             var binder = new NameBindingVisitor(uniqueDeclarations, bindings, builtinNamespaces);
-            binder.Visit(this.ProgramSyntax);
+            binder.Visit(syntaxTree.ProgramSyntax);
 
-            var shortestCycleBySyntax = CyclicCheckVisitor.FindCycles(this.ProgramSyntax, uniqueDeclarations, bindings);
-            foreach (var kvp in shortestCycleBySyntax)
+            var shortestCycleBySymbol = CyclicCheckVisitor.FindCycles(syntaxTree.ProgramSyntax, uniqueDeclarations, bindings);
+            foreach (var kvp in shortestCycleBySymbol)
             {
-                cyclesBySyntax.Add(kvp.Key, kvp.Value);
+                cyclesBySymbol.Add(kvp.Key, kvp.Value);
             }
 
-            // name binding is done
-            // allow type queries now
-            symbolContext.Unlock();
-
-            // TODO: Avoid looping 4 times?
-            var file = new FileSymbol("main",
-                this.ProgramSyntax,
+            // TODO: Avoid looping 5 times?
+            var file = new FileSymbol(
+                syntaxTree.FilePath,
+                syntaxTree.ProgramSyntax,
                 builtinNamespaces,
                 declarations.OfType<ParameterSymbol>(),
                 declarations.OfType<VariableSymbol>(),
                 declarations.OfType<ResourceSymbol>(),
+                declarations.OfType<ModuleSymbol>(),
                 declarations.OfType<OutputSymbol>());
 
+            // name binding is done
+            // allow type queries now
+            symbolContext.Unlock();
+
             return new SemanticModel(file, symbolContext.TypeManager, bindings);
         }
+
+        public bool EmitDiagnosticsAndCheckSuccess(Action<SyntaxTree, Diagnostic> onDiagnostic)
```

Comment:
This is an odd contract. Why not just have a `GetAllDiagnostics()` method that basically does a `SelectMany` (or equivalent lazy-enumerated thing) ordered by file ascending and then diagnostic position ascending. The caller can figure out if this is successful or not. (In CLI, for example we can simply log the diagnostic and the logger should keep track if it had logged an error. This would avoid the allocation of a massive list.)

Reply:
Yeah, I've been meaning to revisit this. It feels like it should be the job of the `Compilation` class to tell you whether compilation is successful, but this contract is definitely wonky.